### DISCUSSION TITLE
regex record under trusted site not working

### DISCUSSION THREAD

**Discussion by bboysnick5** [uBlockOrigin/uAssets#17060] [Q&A]
### Prerequisites

- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uAssets/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] I am running the latest version of uBlock Origin (uBO).

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

My goal is to whitelist any static sites like gstatic.com when 3p dynamic filtering is in play. Below regex does the job when putting into trusted site. However it's not working when 3p dynamic filtering is in use. 

 /^([a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+$/

### URL(s) where the issue occurs.

 /^([a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+$/

### uBO version

latest

### Browser name and version

chrome 110

### Settings

trusted site
hard mode

**Comment by Yuki2718** [comment]
Issues is for issues, do not burden volunteers for your personal question.

**Comment by bboysnick5** [comment]
it's not a question. 

I followed the documents carefully as I understand it's possible to use regex for trusted rules and it overrides both static and dynamic filters. I tested my regex and it works. So why it's not whitelisting the pattern maching sites when 3p dynamic filtering is enabled?

If you think it's not an issue. Please correct my understanding.

**Reply by garry-ut99** [reply]
> Below regex does the job when putting into trusted site.
> I tested my regex and it works

No, it doesn't:
- because it matches only a domain: `gstatic.com`
- but unlike other rules, a regex rule must match a request/URL: `http://www.gstatic.com/`

You:
- didn't actually test it in uBO, but in some regex-tester website 
- or it was working initially but you then altered it right after testing in uBO without further testing in uBO again

This works: 
- `/^https?:\/\/(?:[a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+\//` 
- by the way a little polishing: 
   - a capturing group `(` ==> a non-capturing group `(?:`
   - unescaped dots `.` ==> escaped dots `\.`
- also can add `-` ==> `[a-zA-Z0-9-]` (except the last instance) 
if need to cover cases like `https://abc-static.com/`

**Comment by gorhill** [comment]
URL address bar doesn't show all characters forming the actual URL of the site you are visiting. Copy-paste the URL in the address bar into a text editor, and use the result to craft you regex.

**Reply by garry-ut99** [reply]
Yes, true for OP and Chrome users, also true for Chrome Mobile and Firefox Mobile as well, but what is interesting is that Firefox Desktop displays a full URL (unless it's so long that it's right part doesn't fit into address bar, but in this specific case it would not be a problem for a Firefox Desktop user because most likely it's about a left part [scheme + host] only).

**Reply by gorhill** [reply]
Firefox does not show the trailing `/`, the regex assumed there was nothing beyond the hostname part.

**Reply by garry-ut99** [reply]
> Firefox does not show the trailing `/`

True.

> , the regex assumed there was nothing beyond the hostname part.

Yes, OP's regex incorrectly assumed so, mine didn't, as mine matches full URLs, including a scheme and any path (will match in uBO URLs with anything after first `/`)

As for a reason of incorrect assumption of OP's regex:
- I was initially supposing that OP was trying to create his regex wrongly basing on [the blocklist regex example](https://github.com/gorhill/uBlock/wiki/How-to-mark-a-web-site-as-trusted#disabling-filtering-by-default) which could misled him to a wrong assumption that a regex in uBO needs to match only a hostname and nothing more, and will still catch any URL, just like other rules
- but your point about full URL not being showed in browser's address bar, as a reason of OP's mistake, makes sense as well
- so now I'm not sure anymore which of the reasons was the case for OP's mistake

**Reply by bboysnick5** [reply]
Thanks for the answers and advises. I tinkered around for like https as well as cdns and also you suggestion to accomodate dash. 

However, even if the answer you posted `/^https?:\/\/(?:[a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+\//` does not work on either firefox or chrome on mac. I just turned on all the 3p realted and nothing else.  

I used reddit 'view entired discussion' button to test. Very simple. 

Can you try to reproduce? Anything I missed?

**Reply by stephenhawk8054** [reply]
`/^https?:\/\/(?:[a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+\//`

@bboysnick5 Where did you put this in? Can you screenshot it?

**Comment by bboysnick5** [comment]
![1](https://user-images.githubusercontent.com/16312136/223346038-c88c99cf-79ae-4589-8596-258e21263bbb.png)

**Reply by stephenhawk8054** [reply]
You are using [Trusted sites](https://github.com/gorhill/uBlock/wiki/Dashboard:-Trusted-sites) which is to disable uBO for the page you open in new tab/window, _**NOT**_ for those as network resources in other pages.

I was so confused when reading your post at the first time because `Dynamic filtering` and `Trusted sites` are for different purposes (that's why you see `site` here instead of other terms like network resources...).

What I understand from your post is you want unblock the domain that has `static` while blocking 3rd party scripts using "Dynamic filtering" in [My rules pane](https://github.com/gorhill/uBlock/wiki/Dashboard:-My-rules). Is it correct?

If so, I don't think you can use regex in dynamic filtering rules, and I don't think you can achieve your goal with dynamic filtering, since it's for fast comparing the [hostnames](https://github.com/gorhill/uBlock/wiki/Dynamic-filtering:-rule-syntax) or [URL strings starts with](https://github.com/gorhill/uBlock/wiki/Dynamic-filtering:-rule-syntax), not for comparing a part of the hostname string. Or is it possible @gorhill ?

**Reply by garry-ut99** [reply]
> However, even if the answer you posted `/^https?:\/\/(?:[a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+\//` does not work on either firefox or chrome on mac. I just turned on all the 3p realted and nothing else.
> 
> I used reddit 'view entired discussion' button to test. Very simple.
> 
> Can you try to reproduce? Anything I missed?

The rule on the screenshoot is not the same as the one I proposed, my begins with: `/^https?:` which matches `http` and `https`, while the one on the screenshoot is altered, as it begins with `/^https:` (doesn't contain `?`) which matches `https` only, and not `http`, which can cause the rule to not work if testing with `http` URLS. I don't know whether skipping `?` by OP was intentional or a mistake, and whether OP tested on `http` or `https` URL to claim the rule "does not work".

<hr>

> You are using [Trusted sites](https://github.com/gorhill/uBlock/wiki/Dashboard:-Trusted-sites) which is to disable uBO for the page you open in new tab/window, NOT for those as network resources in other pages.

Yes, clicking 'view entired discussion' changes URL cosmetically, the following explanation is from `removeparam` option which is about Static network filtering, but it applies for Trusted sites & Dynamic network filtering as well:

> https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#removeparam : Cosmetically added params cannot be removed via removeparam (see related comment: [760#issuecomment-724703650](https://github.com/uBlockOrigin/uBlock-issues/issues/760#issuecomment-724703650) and invalid issues: [#1704](https://github.com/uBlockOrigin/uBlock-issues/issues/1704), [#1767](https://github.com/uBlockOrigin/uBlock-issues/issues/1767), [#1951](https://github.com/uBlockOrigin/uBlock-issues/issues/1951), [#2498](https://github.com/uBlockOrigin/uBlock-issues/issues/2498))

**Reply by bboysnick5** [reply]
It was not intentional. I was tinkering around like I said as I don't want to trust http static sites due to clear text. And i tested on reddit which is https://www.redditstatic.com/ and it's https not http. I should have pasted your original regex. 

View entrie discussion is a bad example. Let's try this side window below which will disappear when static is filtered out. 

So apart from my mistakes and carelessness, can I achive what I want?

https://user-images.githubusercontent.com/16312136/223378744-9814c18f-2dff-4b82-8a35-7ccdaf62ed1d.png

**Reply by garry-ut99** [reply]
> So apart from my mistakes and carelessness, can I achive what I want?

Not by Dynamic Filtering, as it  doesn't support regex like already mentioned by steven: https://github.com/uBlockOrigin/uAssets/discussions/17060#discussioncomment-5224329
Thus you need to use Static Filtering like already mentioned by steven: https://github.com/uBlockOrigin/uAssets/discussions/17060#discussioncomment-5224421
Try following examples given by krystian: https://github.com/uBlockOrigin/uAssets/discussions/17060#discussioncomment-5225574

**Comment by bboysnick5** [comment]
> What I understand from your post is you want unblock the domain that has static while blocking 3rd party scripts using "Dynamic filtering" in [My rules pane](https://github.com/gorhill/uBlock/wiki/Dashboard:-My-rules). Is it correct?

Yes I just wanted to have 3p dynamic filtering enabled while noop for static or cdn like sites. I read the doc that regex does not work for dynamic filtering and that trusted sites take precedence over dynamic and static filtering. Hence came up with the workaround. 

Is this possible with the current version? It just turned from an issue to a discussion to a feature request.🤣

**Reply by stephenhawk8054** [reply]
I don't think that's possible, because what I understand from the purpose and implementation of dynamic filtering is for **_fast_** actions over the networks (which are hostnames and URLs that start with).

What you want to achieve needs a more complex structure that's built for [static filtering](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax), implementing it defeats the goal of dynamic filtering. The best way you can do is nooping the domains you want (`gstatic.com`, `hstatic.net`...)

**Reply by bboysnick5** [reply]
yeah albeit tedious...

**Reply by krystian3w** [reply]
Why reddit need unlock gstatic?

**Reply by bboysnick5** [reply]
Reddit has redditstatic.com. The same idea applies.

**Reply by krystian3w** [reply]
Test in the my filters ~ you can add this with domains:

```adb
! allow "static" - recomended add domains 1 by 1 instead whole allow on Internet
@@/^https?:\/\/(?:[a-zA-Z0-9]+\.)?[a-zA-Z0-9]+static\.[a-zA-Z0-9]+\//$domain=reddit.com
```

For overwriting logic, you can then reach for important:
```adb
! disallow specific "static" - based on long url
static.com/hidden-ad*$important,domain=reddit.com
```

> https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#important

if domain with request/resource will be unique then can by blocked in "my rules" (Firewall).

<ins>**But then you have to remember to maintain it when it starts generating unlocked ads/privacy leaks/anti-adblock messages.**</ins>
